Auto-Sync Mips

wargio opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've documented or updated the documentation of every API function and struct this PR changes.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**
Adds options for:
- `CS_OPT_SYNTAX_NOREGNAME` on Mips
- `CS_OPT_SYNTAX_NO_DOLLAR` (new, removes `$` from the register)

Adds support for:
- microMips32r3
- microMips32r6
- nanoMIPS (Generic, NMS1, i7200)
- Mips16
- Mips I ISA
- Mips II ISA
- Mips32 r2 ISA
- Mips32 r3 ISA
- Mips32 r5 ISA
- Mips32 r6 ISA
- Mips III ISA
- Mips IV ISA
- Mips V ISA
- Mips64 r2 ISA
- Mips64 r3 ISA
- Mips64 r5 ISA
- Mips64 r6 ISA
- Octeon (cnMIPS)
- Octeon+ (cnMIPS+)

**cstool** has been refactored for better UX: now by adding `+<feature>` you can directly modify the output of capstone using all options.
```
Arch specific options:
        +att             ATT syntax (only: x86)
        +intel           Intel syntax (only: x86)
        +masm            Intel MASM syntax (only: x86)
        +noregname       Number only registers (only: Arm64, ARM, LoongArch, Mips, PowerPC)
        +moto            Use $ as hex prefix (only: MOS65XX)
        +regalias        Use register aliases, like r9 > sb (only: ARM, Arm64)
        +percentage      Adds % in front of the registers (only: PowerPC)
        +nodollar        Removes $ in front of the registers (only: Mips)
        +nofloat         Disables floating point support (only: Mips)
        +ptr64           Enables 64-bit pointers support (only: Mips)
```

For example:
```
$ ./build/cstool -s mipsel64+nofloat "32 C0 38 46 32 02 20 46 32 03 20 "
 0  32 c0 38 46  .byte  0x32, 0xc0, 0x38, 0x46
 4  32 02 20 46  .byte  0x32, 0x02, 0x20, 0x46
$ ./build/cstool -s mipsel64 "32 C0 38 46 32 02 20 46 32 03 20 "
 0  32 c0 38 46  c.eq.d $f24, $f24
 4  32 02 20 46  c.eq.d $fcc2, $f0, $f0
```

Also fixes the following issues:

- Fix #1519
- Fix #1743
- Fix #1869
- Fix #1990
- Fix #1958

- Fix #2448 
```
$ ./build/cstool mipsel 0400000c
 0  04 00 00 0c  jal    0x10
```

- Fix #1054

```
$ ./build/cstool mips3 0120102d0080e82d40ab500000a0402d0080502d01a0702d40ac1000
 0  01 20 10 2d  move   $v0, $t1
 4  00 80 e8 2d  move   $sp, $a0
 8  40 ab 50 00  dmtc0  $t3, $10, 0
 c  00 a0 40 2d  move   $t0, $a1
10  00 80 50 2d  move   $t2, $a0
14  01 a0 70 2d  move   $t6, $t5
18  40 ac 10 00  dmtc0  $t4, $2, 0
```

- Fix #1133
```
$ ./build/cstool mipsel32r6 b506ff7d
 0  b5 06 ff 7d  pref   0x1f, 0xd($t7)
```

- Fix #1267
```
$ ./build/cstool mips3 00c0502d
 0  00 c0 50 2d  move   $t2, $a2
```

- Fix #1508
```
$ ./build/cstool mips64r2 40021000
 0  40 02 10 00  mfc0   $v0, $2, 0
```

- Fix #1634
```
$ ./build/cstool mips3 46200937
 0  46 20 09 37  c.ule.d        $fcc1, $f1, $f0
```

- Fix #1673
```
$ ./build/cstool mipsel64 03800c4080008c35
 0  03 80 0c 40  mfc0   $t4, $16, 3
 4  80 00 8c 35  ori    $t4, $t4, 0x80
```

- Fix #1680
```
$ ./build/cstool -d mipsel 4000000c0800e003
 0  40 00 00 0c  jal    0x100
        ID: 34 (jal)
        op_count: 1
                operands[0].type: IMM = 0x100
        Groups: call HasStdEnc NotInMicroMips 

 4  08 00 e0 03  jr     $ra
        ID: 773 (jr)
        op_count: 1
                operands[0].type: REG = ra
        Groups: return jump HasStdEnc NotMips32r6 NotMips64r6 NotInMicroMips 
```

- Fix #1780
```
# the test in the example is always decoded as xori, i checked also with other disassemblers.
$ ./build/cstool mips 7c03e83b
 0  7c 03 e8 3b  rdhwr  $v1, $29
```

- Fix #1851
```
$ ./build/cstool mipsel64 "32 C0 38 46 32 02 20 46 32 03 20 46"
 0  32 c0 38 46  c.eq.d $f24, $f24
 4  32 02 20 46  c.eq.d $fcc2, $f0, $f0
 8  32 03 20 46  c.eq.d $fcc3, $f0, $f0
```

- Fix #1957
```
$ ./build/cstool mipsel32r6 "000019f8"
 0  00 00 19 f8  jalrc  $t9
```

wargio committed:
Update MIPS via auto-sync from latest LLVM

Adds support for:
- microMips32r3
- microMips32r6
- Mips16
- Mips I ISA
- Mips II ISA
- Mips32 r2 ISA
- Mips32 r3 ISA
- Mips32 r5 ISA
- Mips32 r6 ISA
- Mips III ISA
- Mips IV ISA
- Mips V ISA
- Mips64 r2 ISA
- Mips64 r3 ISA
- Mips64 r5 ISA
- Mips64 r6 ISA
- Octeon (cnMIPS)
- Octeon+ (cnMIPS+)

Also cleans up cstool for better output and support.

Add mips into auto-sync.yml

Add mips details

Add reg_access

Fix mapping for jr $ra which is not a return type when it should be.

Remove save-restore points in mips

Implement CS_OPT_SYNTAX_NOREGNAME and CS_OPT_SYNTAX_NO_DOLLAR

Rot127 commented:
The auto-sync test is fixed in the `modern_testing` PR.

Rot127 commented on arch/Mips/MipsInstPrinter.c:
```diff
@@ -269,10 +278,12 @@ static void printMemOperand(MCInst *MI, int opNum, SStream *O)
 	SStream_concat0(O, "(");
 	printOperand((MCInst *)MI, opNum, O);
 	SStream_concat0(O, ")");
+	Mips_set_mem_access(MI, false);
```
As discussed

wargio replied:
i have modified this following what also the other archs does (they implement `set_mem_access`).

Rot127 reviewed (changes requested):
As discussed move details to MipsMapping.c

Rot127 committed:
Move patch constraints into the config file and add a test (#2)

* Move patch constraints into the config file and add a test
* Handle some Mips operand printer.

wargio committed:
Add nanoMIPS & refactor cstool

mtk-pub/nanomips-llvm16: 86653206c89a317ba2764d5d7dc9f31afefd2a63
date: Sat Aug 10 10:26:29 2024 +0200

Rot127 commented on arch/Mips/MipsInstPrinter.c:
```diff
@@ -1,261 +1,263 @@
-		MCInst_setOpcodePub(MI, Mips_map_insn(mnem));
-		cs_mem_free(mnem);
-	}
+void Mips_LLVM_printInst(MCInst *MI, uint64_t Address, SStream *O) {
+	if (!printAliasInstr(MI, Address, O) && !printAlias4(MI, Address, O))
+		printInstruction(MI, Address, O);
 }
```
Please handle alias details here. With v6 we add support to select the alias and real details.

You can check how it is done in `ARMInstPrinter.c::printInst()`. Just delete the big `switch` statement in there and the logic should be clear.

This will be essential for Rizin as well. Because you can always get the real instruction details of any instruction. And greatly simplify the RzIL code for it! No need anymore to implement a RzIL operation for an alias and the (effectively the same) real instruction.
It is also important because I'd like to allow in the future to select any combination of (non)-alias syntax and (non)-alias details.

Also add a test case. Just to see that it gets decoded correctly.

wargio replied:
this is the original `printInst` without the switch.

Rot127 replied:
I mean these lines:
https://github.com/capstone-engine/capstone/blob/df7228674963b8bc0fe6e00c29e13c731a665d41/arch/ARM/ARMInstPrinter.c#L110-L112
https://github.com/capstone-engine/capstone/blob/df7228674963b8bc0fe6e00c29e13c731a665d41/arch/ARM/ARMInstPrinter.c#L419-L432

Rot127 replied:
This is added by us. To close the stream and not print the asm text, but still call the `add_cs_detail` code to add the operands. If the "real" details are requested. But the instruction is an alias. This way we get "alias asm-text" but "real-details"

Rot127 commented on arch/Mips/MipsInstPrinter.c:
```diff
@@ -265,160 +267,348 @@ static void printMemOperand(MCInst *MI, int opNum, SStream *O)
 	// when using stack locations for not load/store instructions
 	// print the same way as all normal 3 operand instructions.
-	printOperand(MI, opNum, O);
+	printOperand((MCInst *)MI, opNum, O);
```
These casts are an artifact from generation? I can't find them in the cpp code.

wargio replied:
no, these are to make the compiler happy when moving from `const MCInst *` to `MCInst *`, maybe in `printMemOperandEA` is useless, but in other areas is.

Rot127 commented on arch/Mips/MipsInstPrinter.c:
```diff
@@ -265,160 +267,348 @@ static void printMemOperand(MCInst *MI, int opNum, SStream *O)
 {
-	MCOperand *MO = MCInst_getOperand(MI, opNum);
-	SStream_concat0(O, MipsFCCToString((Mips_CondCode)MCOperand_getImm(MO)));
+	MCOperand *MO = MCInst_getOperand(MI, (opNum));
```
Missing `add_cs_detail()`?

wargio replied:
no, the fcc operand is a string to determine the opcode name. its constructed like `c.%s.d` where `%s` is something like `cmp` or `lt` etc.. and returned by MipsFCCToString
these are already defined by the opcode number

Rot127 replied:
I see. If it turns out useful to have this information as an extra field in the details, we can also add it later.

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
-	{ MIPS_GRP_MIPS4_32, "mips4_32" },
-	{ MIPS_GRP_MIPS4_32R2, "mips4_32r2" },
-	{ MIPS_GRP_MIPS5_32R2, "mips5_32r2" },
+	assert((map_get_op_type(MI, OpNum) & ~CS_OP_MEM) == CS_OP_REG);
```
```suggestion
	CS_ASSERT((map_get_op_type(MI, OpNum) & ~CS_OP_MEM) == CS_OP_REG);
```

Also below.

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
+	cs_op_type op_type = map_get_op_type(MI, OpNum) & ~CS_OP_MEM;
+	if (op_type == CS_OP_IMM) {
+		uint64_t Target = (uint64_t)MCInst_getOpVal(MI, OpNum);
+		Mips_set_detail_op_imm(MI, OpNum, Target + MI->address);
```
You can check for the `csh->PrintBranchImmNotAsAddress` option here and only optionally add the address.

Also please rename the `CS_OPT_NO_BRANCH_OFFSET` option (it sets this flag). The naming was horrible chosen by me.

Rot127 replied:
Ouh, but the immediate offset is a `int64_t`. Not an `uint64_t`!. The LLVM code does:
```cpp
    uint64_t Target = Address + Op.getImm();
```

But `getImm(Op)` returns an `int64_t` (also our C version).

wargio replied:
this is a copy of what the printing instructions do. otherwise i need to use args to get this value (`address`).

Rot127 replied:
Sorry, I got confused while thinking about it. It doesn't matter of course, both are 64bit

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
+static void Mips_set_detail_op_uimm(MCInst *MI, unsigned OpNum)
+{
+	Mips_set_detail_op_imm(MI, OpNum,
+				    MCInst_getOpVal(MI, OpNum));
```
Are `uint64_t` immediate values possible here? Because `Mips_set_detail_op_imm` would cast them silently to `int64_t`

Rot127 replied:
If yes, I would simply add a flag to `Mips_set_detail_op_imm` to indicate if it is an unsigned/signed type.

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
+
+static void Mips_set_detail_op_mem_nanomips(MCInst *MI, unsigned OpNum)
+{
+	if (!detail_is_set(MI) || !doing_mem(MI))
```
Better move the `doing_mem()` check to a separated `if` and print a warning if it is false. The function should only be called for memory operands. Not returning silently will ease debugging and discovering bugs a lot.

Rot127 replied:
Especially discovering faulty logic after this big refactor.

wargio replied:
that `doing_mem` should be always false. it's there just in case somebody modifies the code. it could be an assert.

Rot127 replied:
`CS_ASSERT` is better in my experience (gets removed in release build). I spent quite some debugging time in the past, because these things were silent.

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
-	i = name2id(&insn_name_maps[1], ARR_SIZE(insn_name_maps) - 1, name);
+	if (isNanoMips) {
+		for (unsigned i = OpNum; i < MCInst_getNumOperands(MI); i++) {
+			Mips_set_detail_op_reg(MI, i, MCInst_getOpVal(MI, i));
```
Please add an additional `part_of_reglist` (or similar) flag to the `mips_op`. It is pretty essential information. Because instructions with a register list and regs outside of lists, become indistinguishable otherwise.

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
+	}
+	// -2 because register List is always first operand of instruction
+	// and it is always followed by memory operand (base + offset).
+	for (unsigned i = OpNum, e = MCInst_getNumOperands(MI) - 2; i != e; ++i) {
```
Please add an underflow check with `CS_ASSERT()` for the `-2`.

wargio replied:
can't be. it's checked by the code which calls this.

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
-	return (i != -1)? i : MIPS_REG_INVALID;
+static void Mips_set_detail_op_uimm_address(MCInst *MI, unsigned OpNum)
+{
+	uint64_t Target = MI->address + (uint64_t)MCInst_getOpVal(MI, OpNum);
```
Same as above. Check if address should be added at all.

wargio replied:
this is a copy of what the printing instructions do. otherwise i need to use args to get this value (`address`).

Rot127 replied:
See above. Didn't think it to an end properly.

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,454 @@
+static void Mips_set_detail_op_uimm_address(MCInst *MI, unsigned OpNum)
+{
+	uint64_t Target = MI->address + (uint64_t)MCInst_getOpVal(MI, OpNum);
+	Mips_set_detail_op_imm(MI, OpNum, Target);
```
~~As above. Just check if typing is ok in this case.~~

Rot127 commented on arch/Mips/MipsMapping.h:
```diff
@@ -1,25 +1,61 @@
-mips_reg Mips_map_register(unsigned int r);
+static inline void set_mem_access(MCInst *MI, bool status)
+{
+	if (!MI->flat_insn->detail)
```
```suggestion
	if (!detail_is_set(MI))
```

Rot127 commented on cstool/cstool.c:
```diff
@@ -663,19 +685,27 @@ int main(int argc, char **argv)
+		if (len == mode_len && !strncmp(all_archs[i].name, mode, mode_len)) {
 			arch = all_archs[i].arch;
-			err = cs_open(all_archs[i].arch, all_archs[i].mode, &handle);
+			md = all_archs[i].mode;
```
```suggestion
			mode = all_archs[i].mode;
```
And above `mode_arg`/`mode_arg_len` or something.
No need to continue with the imprecise naming from before :)

Rot127 commented on cstool/cstool.c:
```diff
@@ -688,7 +718,8 @@ int main(int argc, char **argv)
 	if (err) {
-		printf("ERROR: Failed on cs_open(), quit!\n");
+		const char *error = cs_strerror(err);
+		printf("ERROR: Failed on cs_open(): %s\n", error);
```
We should probably print to `stderr` but this is for another PR.

Rot127 reviewed (changes requested):
Thanks a lot! Awesome job! This was a big one. And it looks great.
Please address the comments.

Also, in the long run we can think about adding the instruction formats as additional information.

LLVM defines them well on a shallow look (the ISA doesn't though). But I like to have them provide them, because if Capstone fails badly to provide certain details, people can extract them by themselves until there is a fix.
Of course referencing `MipsInstrFormats.td`.

Also please rebase onto https://github.com/capstone-engine/capstone/pull/2456 and generate the yaml tests. The current MC tests can stay as they are. Don't delete them. They are still used by the fuzzer.

wargio committed:
Optimize tests and fix imm in map & null deref due invalid pointer.

wargio committed:
Updated bindings for mips.

Rot127 commented on suite/cstest/src/test_run.c:
```diff
@@ -102,17 +102,19 @@ static bool parse_input_options(const TestInput *input, cs_arch *arch,
-		if (mode_found) {
-			*mode |= val;
-			goto next_option;
+		for (size_t k = 0; k < ARR_SIZE(test_mode_map); k++) {
```
Please introduce the binary search again. We get a worst case number of string comparisons of `O(log2(150) = 7) * ~3 options (assuming there are on average 3 options per test case).
Vs. O(150) * 3 string comparisons.

Rot127 replied:
You can sort the table as documented in `Mapping.h::cs_enum_id_map`.

wargio replied:
that function is broken or similar. never worked correctly, why? dunno. that is why i have removed it.

Rot127 commented on suite/cstest/src/main.c:
Remove empty file.

Rot127 commented on suite/auto-sync/src/autosync/cpptranslator/saved_patches.json:
You can update the file. Just run `ASUpdater -s IncGen Traslate Diff -a Mips` and press `O` (capital "O") for all decisions. This will select always the version in the current `Mips**` source files in `arch/Mips/`. It will help later, when we update. Because even if the Differ is not nice, it automatically applies the patches.

Rot127 commented on include/capstone/mips.h:
```diff
@@ -244,6 +687,12 @@ typedef struct cs_mips_op {
 		mips_op_mem mem;	///< base/index/scale/disp value for MEM operand
 	};
+	bool is_reglist;    ///< defines if the register is part of a list
+	bool is_unsigned;   ///< when true, the immediate value is unsigned
```
Are there `uint64_t` possible in Mips? If yes, the immediate value above should be of the type. Otherwise this is fine.

Rot127 replied:
Also see comment below.

wargio replied:
there is no difference for capstone since type IMM is always int64, so what matters?

Rot127 commented on arch/Mips/MipsMapping.c:
```diff
@@ -1,1071 +1,467 @@
-	{ MIPS_GRP_MIPS3_32, "mips3_32"},
-	{ MIPS_GRP_MIPS3_32R2, "mips3_32r2" },
+	Mips_get_detail_op(MI, 0)->type = MIPS_OP_IMM;
+	Mips_get_detail_op(MI, 0)->imm = (int64_t)Imm;
+	Mips_get_detail_op(MI, 0)->is_unsigned = true;
```
As said above. If `uint64` values exist, you have the choice between changing the API here (and make it a `uint64_t`) or document it in code and the release document. In any case it just needs to be clear who the field should be interpreted. By bits or by type.

When you refactor Rizin and see that one interpretation is vastly better, you can still change it. Preferable though before Alpha and at latest before the Beta release.

Rot127 commented on arch/Mips/MipsInstPrinter.c:
```diff
@@ -1,261 +1,280 @@
-		case Mips_Restore16:
-		case Mips_RestoreX16:
-			return;
+void Mips_LLVM_printInst(MCInst *MI, uint64_t Address, SStream *O) {
+	bool useAliasDetails = map_use_alias_details(MI);
+	if (!useAliasDetails) {
+		SStream_Close(O);
+		printInstruction(MI, Address, O);
+		SStream_Open(O);
+		map_set_fill_detail_ops(MI, false);
 	}
```
Please add at least one test for the "alias instruction with real details". Just to make sure it works as expected.

Rot127 reviewed (changes requested):
Ok nice. Almost done. Some more nitpicks:

- Please add the resolved issues (from the PR descriptions above) as tests to `tests/issues.yaml`.
- Please add the newly added features (`access`, `is_reglist` etc.) and new restrictions (checks features more strictly) for Mips in `docs/cs_v6_release_guide.md`.
- Also document new option (`CS_OPT_SYNTAX_NO_DOLLAR`) and `cstool` behavior there.

wargio committed:
Fix requested in comments.

wargio committed:
Add changes to v6 docs

Rot127 commented on include/capstone/mips.h:
```diff
@@ -684,7 +684,8 @@ typedef struct cs_mips_op {
 		mips_reg reg;		///< register id for REG operand
-		int64_t imm;		///< immediate value for IMM operand
+		int64_t imm;		///< signed immediate value for IMM operand
+		int64_t uimm;		///< unsigned immediate value for IMM operand
```
`uint64_t uimm;`, right?
Also document for `is_unsigned` that `uimm` instead of `imm` is set if true.

wargio replied:
it's documented, i mean, it's obvious, no?

Rot127 commented:
```diff
diff --git a/suite/auto-sync/src/autosync/cpptranslator/patches/FieldFromInstr.py b/suite/auto-sync/src/autosync/cpptranslator/patches/FieldFromInstr.py
index 67f7faae..58092ea0 100644
--- a/suite/auto-sync/src/autosync/cpptranslator/patches/FieldFromInstr.py
+++ b/suite/auto-sync/src/autosync/cpptranslator/patches/FieldFromInstr.py
@@ -41,9 +41,9 @@ class FieldFromInstr(Patch):
 
         # Determine width of instruction by the variable name.
         if ffi_first_arg_text[-2:] == "32":
-            inst_width = 4
+            inst_width = b"4"
         elif ffi_first_arg_text[-2:] == "16":
-            inst_width = 2
+            inst_width = b"2"
         else:
             # Get the Val/Inst parameter.
             # Its type determines the instruction width.
diff --git a/.github/workflows/auto-sync.yml b/.github/workflows/auto-sync.yml
index 9d26470d..d1428d06 100644
--- a/.github/workflows/auto-sync.yml
+++ b/.github/workflows/auto-sync.yml
@@ -101,4 +101,3 @@ jobs:
           ./src/autosync/cpptranslator/Differ.py -a ARM --check_saved
           ./src/autosync/cpptranslator/Differ.py -a PPC --check_saved
           ./src/autosync/cpptranslator/Differ.py -a LoongArch --check_saved
-          ./src/autosync/cpptranslator/Differ.py -a Mips --check_saved


```

wargio commented:
@Rot127 fixed

Rot127 reviewed (approved):
Excellent!

wargio commented:
@kabeor

XVilka commented:
Would be nice to get it merged, so Coverity would check this code too, also we could update it in Rizin @kabeor

kabeor reviewed (approved):
Really awesome, thank you!
